### PR TITLE
[BC Break] Make data providers for abstract test cases static

### PR DESCRIPTION
**Pull request by OskarStark**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Refs #48668 
| License       | MIT
| Doc PR        | -

Follows https://github.com/symfony/symfony/pull/48668#issuecomment-1429413004

### Todos
* [x] Bump min version in `composer.json` Notifier @OskarStark 
* [x] Bump min version in `composer.json` Translation @OskarStark 
* [x] Bump min version in `composer.json` Mailer @OskarStark 
* [x] Bump min version in `composer.json` Security @OskarStark - _nothing todo_
* [x] document BC break in CHANGELOG files Notifier @OskarStark
* [x] document BC break in CHANGELOG files Translation @OskarStark
* [x] document BC break in CHANGELOG files Mailer @OskarStark
* [x] document BC break in CHANGELOG files Security @OskarStark
* [x] document BC break in UPGRADE files Notifier @OskarStark
* [x] document BC break in UPGRADE files Translation @OskarStark
* [x] document BC break in UPGRADE files Mailer @OskarStark
* [x] document BC break in UPGRADE files Security @OskarStark
* [x] remove $this in static methods @alexandre-daubois 
* [x] make tests green

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Todo:
- remove $this in static methods
- bump composer.json
- make tests green
- document BC break in CHANGELOG + UPGRADE files

File: src/Symfony/Component/Notifier/Test/TransportFactoryTestCase.php [hunk 1]
```diff
@@ -17,6 +17,8 @@
 use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
 use Symfony\Component\Notifier\Transport\Dsn;
 use Symfony\Component\Notifier\Transport\TransportFactoryInterface;
+use Symfony\Component\Notifier\Transport\TransportInterface;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Test/TransportFactoryTestCase.php]: Missing use

File: tools/rector.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
```
Comment (nicolas-grekas) [inline tools/rector.php]: should be removed

File: src/Symfony/Component/Translation/Test/ProviderTestCase.php [hunk 3]
```diff
@@ -37,9 +37,9 @@ abstract class ProviderTestCase extends TestCase
     abstract public function createProvider(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, string $defaultLocale, string $endpoint): ProviderInterface;
 
     /**
-     * @return iterable<array{0: string, 1: ProviderInterface}>
+     * @return iterable<array{0: ProviderInterface, 1: string}>
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Translation/Test/ProviderTestCase.php:42]: rebase needed

File: tools/rector.php (hunk 2 above)
Comment (OskarStark) [reply tools/rector.php]: Done

**Comment by OskarStark** [comment]
Rebased

**Comment by nicolas-grekas** [comment]
(rebase unlocked)

**Comment by alexandre-daubois** [comment]
Notifier PR against your branch as you asked is ready at https://github.com/OskarStark/symfony/pull/1 👍

**Comment by OskarStark** [comment]
Rebased and merged https://github.com/OskarStark/symfony/pull/1 👍

File: src/Symfony/Component/Security/Core/CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.4.21
+------
+
+* [BC BREAK] `AccessDecisionStrategyTestCase::provideStrategyTests()` is now static
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/CHANGELOG.md]: space

File: UPGRADE-5.4.md [hunk 5]
```diff
@@ -78,8 +86,8 @@ Monolog
 Notifier
 --------
 
- * [BC BREAK] The following data providers for `TransportTestCase` are now static: `toStringProvider()`, `supportedMessagesProvider()` and `unsupportedMessagesProvider()`
- * [BC BREAK] The `TransportTestCase::createTransport()` method is now static
```
Comment (nicolas-grekas) [inline UPGRADE-5.4.md:82]: why removing?

File: src/Symfony/Component/Mailer/CHANGELOG.md [hunk 6]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+5.4.21
+------
+
+* [BC BREAK] The following data providers for `TransportFactoryTestCase` are now static:
+  `supportsProvider()`, `createProvider()`, `unsupportedSchemeProvider()`and `incompleteDsnProvider()`
+* [BC BREAK] The following data providers for `TransportTestCase` are now static:
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/CHANGELOG.md]: space

**Comment by nicolas-grekas** [comment]
please squash when pushing, it'd be great to have only two commits in the PR, one for each of you

**Comment by alexandre-daubois** [comment]
I think this is the final one before having an all-green on 5.4 🙌 

https://github.com/OskarStark/symfony/pull/2

**Comment by OskarStark** [comment]
Merged. Will squash them properly later today

File: UPGRADE-5.4.md (hunk 5 above)
Comment (OskarStark) [reply UPGRADE-5.4.md:82]: I found these occurrences only in Changelog files, but not in Upgrade files.

Add it back?

File: src/Symfony/Component/Mailer/CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+5.4.21
+------
+
+* [BC BREAK] The following data providers for `TransportFactoryTestCase` are now static:
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/CHANGELOG.md]:
```suggestion
 * [BC BREAK] The following data providers for `TransportFactoryTestCase` are now static:
```

File: src/Symfony/Component/Mailer/CHANGELOG.md (hunk 6 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/CHANGELOG.md]:
```suggestion
 * [BC BREAK] The following data providers for `TransportTestCase` are now static:
```

File: src/Symfony/Component/Security/Core/CHANGELOG.md (hunk 4 above)
Comment (OskarStark) [inline src/Symfony/Component/Security/Core/CHANGELOG.md]:
```suggestion
 * [BC BREAK] `AccessDecisionStrategyTestCase::provideStrategyTests()` is now static
```

File: UPGRADE-5.4.md [hunk 8]
```diff
@@ -63,6 +63,14 @@ Lock
  * Deprecate usage of `PdoStore` with a `Doctrine\DBAL\Connection` or a DBAL url, use the new `DoctrineDbalStore` instead
  * Deprecate usage of `PostgreSqlStore` with a `Doctrine\DBAL\Connection` or a DBAL url, use the new `DoctrineDbalPostgreSqlStore` instead
 
+Mailer
+------
+
+ * The following data providers for `TransportFactoryTestCase` are now static:
+   `supportsProvider()`, `createProvider()`, `unsupportedSchemeProvider()`and `incompleteDsnProvider()`
+ * The following data providers for `TransportTestCase` are now static:
+   `toStringProvider()`, `supportedMessagesProvider()` and `unsupportedMessagesProvider()`
```
Comment (OskarStark) [inline UPGRADE-5.4.md]:
```suggestion
 * The following data providers for `TransportFactoryTestCase` are now static: `supportsProvider()`, `createProvider()`, `unsupportedSchemeProvider()`and `incompleteDsnProvider()`
 * The following data providers for `TransportTestCase` are now static: `toStringProvider()`, `supportedMessagesProvider()` and `unsupportedMessagesProvider()`
```

File: UPGRADE-5.4.md [hunk 9]
```diff
@@ -197,3 +204,10 @@ Security
    $token = new PreAuthenticatedToken($user, $firewallName, $roles);
    $token = new SwitchUserToken($user, $firewallName, $roles, $originalToken);
    ```
+
+Translation
+-----------
+
+ * The following data providers for `ProviderFactoryTestCase` are now static:
+   `supportsProvider()`, `createProvider()`, `unsupportedSchemeProvider()`and `incompleteDsnProvider()`
```
Comment (OskarStark) [inline UPGRADE-5.4.md]:
```suggestion
 * The following data providers for `ProviderFactoryTestCase` are now static: `supportsProvider()`, `createProvider()`, `unsupportedSchemeProvider()`and `incompleteDsnProvider()`
```

File: UPGRADE-5.4.md (hunk 5 above)
Comment (nicolas-grekas) [reply UPGRADE-5.4.md:82]: I think we should add the prefix in all files yes, to make this clear

**Comment by OskarStark** [comment]
rebased

**Comment by nicolas-grekas** [comment]
Thank you @OskarStark.

**Pull request merged**